kong/router/compat.lua

```diff
@@ -295,8 +302,58 @@ local function get_exp_and_priority(route)
 end
 
 
-function _M.new(routes, cache, cache_neg, old_router)
-  return atc.new(routes, cache, cache_neg, old_router, get_exp_and_priority)
+-- group array-like table t by the function f, returning a table mapping from
+-- the result of invoking f on one of the elements to the actual elements.
+local function group_by(t, f)
+  local result = {}
+  for _, value in ipairs(t) do
+    local key = f(value)
+    if result[key] then
+      result[key][#result[key] + 1] = value
+    else
+      result[key] = {value}
+    end
+  end
+  return result
+end
+
+-- split routes into multiple routes, one for each prefix length and one for all
+-- regular expressions
+local function split_route_by_path_into(rs, split_rs)
+  if is_empty_field(rs.route.paths) or #rs.route.paths == 1 then
+    split_rs[#split_rs + 1] = rs
+    return
+  end
+
+  assert(rs.route.paths)
+  local grouped_paths = group_by(
+    rs.route.paths,
+    function(path)
+      return is_regex_magic(path) or #path
+    end
```

Comment:
I don't know if luajit will be clever enough to optimize away the closure creation here, but unless I'm mistaken this can just as easily be turned into a regular `local function` at the module scope rather than a closure.

Reply:
There is no difference between the two in my measurements, so I'm leaving it as-is.